src/core/Semantics.ml

```diff
@@ -43,69 +43,6 @@ let get_local_tp i =
   | v -> EvM.ret v
   | exception _ -> EvM.throw @@ NbeFailed "Variable out of bounds"
 
-
-let simplify_cof phi =
-  let open CM in
-  let+ thy = read_cof_thy in
-  CofThy.Disj.simplify_cof thy phi
-
-module FaceLattice :
-sig
-  (** An atomic formula *)
-  type atom = [`CofEq of D.dim * D.dim]
-
-  (** A generator for a lattice homomorphism *)
-  type gen = atom -> D.cof CM.m
-
-  (** Extend a generator to a lattice homomorphism *)
-  val extend : gen -> D.cof -> D.cof CM.m
-
-  (** Quantifier elimination *)
-  val forall : DimProbe.t -> D.cof -> D.cof CM.m
-end =
-struct
-  open CM
-  module K = Kado.Syntax
-
-  type atom = [`CofEq of D.dim * D.dim]
-  type gen = atom -> D.cof CM.m
-
-  let extend gen =
-    let rec loop =
-      function
-      | K.Var _ as phi -> ret phi
-      | K.Cof phi ->
-        match phi with
-        | K.Join psis ->
-          let+ psis = MU.map loop psis in
-          K.Free.join psis
-        | K.Meet psis ->
-          let+ psis = MU.map loop psis in
-          K.Free.join psis
-        | K.Eq (r, s) ->
-          gen @@ `CofEq (r, s)
-    in
-    loop
-
-  let forall sym =
-    let i = Dim.DimProbe sym in
-    extend @@
-    function
```

Comment:
I am still a tiny bit concerned about the invariants here. But I do not have a counterexample... @favonia Can you perhaps explain why you are mostly certain this is OK?

Reply:
I believe I did not change the functionality of the code---that is, it is wrong iff it was wrong. Notice that there are _two_ forall now---one is syntactic (cheap) and one is semantic (expensive), and it is the semantic (expensive) one used here. I _am_ worried about the invariant but I was already worried before this PR. Does this answer your question?

Reply:
Oh cool!! Sorry for my 🌰  🧠

Reply:
> Where is the expensive one located now?

https://github.com/RedPRL/kado/pull/6